### PR TITLE
AgentCLI command to debug detected dependencies

### PR DESCRIPTION
**Pull request by anderruiz**
# What Does This Do

New command line tool to scan dependencies in a jar/war file or in a folder

# Motivation

Provide a tool that gathers more information that what it is stored in the backend to be able to debug weird scenarios

### CODE REVIEW

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentCLI.java [hunk 1]
```diff
@@ -77,4 +85,72 @@ public static void uploadCrash(final String[] args) throws Exception {
...
+                      }
+                    }
+                    log.debug("Adding new jar:" + temp.getAbsolutePath());
+                    recursiveDependencySearch(invoker, temp);
+                    temp.delete();
+                  } catch (Exception ex) {
+                    throw new RuntimeException(ex);
+                  }
+                } else {
+                  log.debug("Entry:" + e.getName() + " ignored in file:" + file.getAbsolutePath());
```
Comment (smola) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentCLI.java:151]:
```suggestion
                  log.debug("Entry: {} ignored in file: {}",  e.getName() , file.getAbsolutePath());
```

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentCLI.java [hunk 2]
```diff
@@ -77,4 +85,72 @@ public static void uploadCrash(final String[] args) throws Exception {
...
+                        "Jar entry found in file:" + file.getName() + " entry:" + e.getName());
+                    File temp = File.createTempFile("internal", ".jar");
+                    try (InputStream is = jar.getInputStream(e);
+                        OutputStream out = new FileOutputStream(temp)) {
+                      int read;
+                      while ((read = is.read()) != -1) {
+                        out.write(read);
+                      }
+                    }
+                    log.debug("Adding new jar:" + temp.getAbsolutePath());
```
Comment (smola) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentCLI.java:144]:
```suggestion
                    log.debug("Adding new jar: {}", temp.getAbsolutePath());
```

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentCLI.java [hunk 3]
```diff
@@ -77,4 +85,72 @@ public static void uploadCrash(final String[] args) throws Exception {
...
+    try (JarFile jar = new JarFile(file)) {
+      log.debug("Finding entries in file:" + file.getName());
+
+      jar.stream()
+          .forEach(
+              e -> {
+                if (e.getName().endsWith(".jar") || e.getName().endsWith(".war")) {
+                  try {
+                    log.debug(
+                        "Jar entry found in file:" + file.getName() + " entry:" + e.getName());
```
Comment (smola) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentCLI.java:135]:
```suggestion
                        "Jar entry found in file: {} entry: {}", file.getName(), e.getName());
```

File: dd-java-agent/src/main/java/datadog/trace/bootstrap/AgentJar.java [hunk 4]
```diff
@@ -30,6 +30,9 @@ public static void main(final String[] args) {
           case "uploadCrash":
             uploadCrash(args);
             break;
+          case "scandependencies":
```
Comment (smola) [inline dd-java-agent/src/main/java/datadog/trace/bootstrap/AgentJar.java:33]:
```suggestion
          case "scanDependencies":
```

To maintain consistency with case style of other commands.

File: dd-java-agent/src/main/java/datadog/trace/bootstrap/AgentJar.java [hunk 5]
```diff
@@ -59,6 +62,7 @@ private static void printUsage() {
     System.out.println("usage:");
     System.out.println("  sampleTrace [-c count] [-i interval]");
     System.out.println("  uploadCrash file ...");
+    System.out.println("  scandependencies file/path ...");
```
Comment (smola) [inline dd-java-agent/src/main/java/datadog/trace/bootstrap/AgentJar.java:65]:
```suggestion
    System.out.println("  scanDependencies <path> ...");
```

File: telemetry/src/main/java/datadog/telemetry/dependency/Dependency.java [hunk 6]
```diff
@@ -239,7 +247,7 @@ public static synchronized Dependency guessFallbackNoPom(
       while (is.read(buf, 0, buf.length) > 0) {}
       hash = String.format("%040X", new BigInteger(1, md.digest()));
     }
-
+    log.debug("No maven dependency added {}.{} jar name {} hash {}");
```
Comment (smola) [inline telemetry/src/main/java/datadog/telemetry/dependency/Dependency.java:250]: There are placeholders here, but no arguments?

**Review by smola — CHANGES_REQUESTED** [review]

**Review by ValentinZakharov — APPROVED** [review]

**Comment by ValentinZakharov** [comment]
Does it make sense turn debug messages for log file, turn into stdout? 🤔 
I'm doubt somebody will use tracer as CLI with enabled debug mode.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentCLI.java [hunk 7]
```diff
@@ -77,4 +85,72 @@ public static void uploadCrash(final String[] args) throws Exception {
...
+                      int read;
+                      while ((read = is.read()) != -1) {
+                        out.write(read);
+                      }
+                    }
+                    log.debug("Adding new jar:" + temp.getAbsolutePath());
+                    recursiveDependencySearch(invoker, temp);
+                    temp.delete();
+                  } catch (Exception ex) {
+                    throw new RuntimeException(ex);
```
Comment (dougqh) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentCLI.java:148]: Why are we propagating a RuntimeException?
What are we doing in the event that delete fails?  It doesn't throw an Exception; it returns a boolean.

File: telemetry/src/main/java/datadog/telemetry/dependency/Dependency.java (hunk 6 above)
Comment (anderruiz) [reply telemetry/src/main/java/datadog/telemetry/dependency/Dependency.java:250]: Done

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentCLI.java (hunk 7 above)
Comment (anderruiz) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentCLI.java]: Changes done

**Comment by anderruiz** [comment]
> Does it make sense turn debug messages for log file, turn into stdout? 🤔 I'm doubt somebody will use tracer as CLI with enabled debug mode.

Part of the normal execution will go to trace so it makes sense to use the same approach

**Review by smola — APPROVED** [review]

**Pull request merged**